### PR TITLE
implemented workspace list

### PR DESCRIPTION
**Pull request by dvoet**
also fixed PerRequest actor to marshall responses correctly

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/WorkspaceDAO.scala [hunk 1]
```diff
@@ -37,4 +43,25 @@ class FileSystemWorkspaceDAO(storageDirectory: Path) extends WorkspaceDAO {
...
   }
+
+  def list(): Seq[WorkspaceShort] = {
+    val visitor = new WorkspaceFileVisitor(storageDirectory)
+    Files.walkFileTree(storageDirectory, visitor)
+    visitor.workspaces.result
+  }
 }
+
+private class WorkspaceFileVisitor(root: Path) extends SimpleFileVisitor[Path] with LazyLogging {
```
Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/WorkspaceDAO.scala]: I feel like whenever you find yourself reaching into GoF there's probably a better way to do things (a la the notion that design patterns make up for missing language features). Like here, it's really just a foreach() across the collection of all files in the dir tree.

So this bothers me but in this particular case I think it's fine since we already have SimpleFileVisitor which already codes up the recursive dir tree generation. But it'd be nice if we could say something like:

     getAllFilesRecursively(path) foreach { doSomething() }

tl;dr - this disturbs my inner FP-fan enough to speak up but you should probably leave it as is

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -23,10 +26,15 @@ object WorkspaceService {
...
+    case ListWorkspaces => list
   }
 
   def saveWorkspace(workspace: Workspace): Unit = {
     workspaceDAO.save(workspace)
-    context.parent ! PerRequest.RequestComplete(StatusCodes.Created, None)
+    context.parent ! PerRequest.RequestCompleteNoContent(StatusCodes.Created)
+  }
+
+  def list() = {
```
Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: why is one func named saveWorkspace and the other just list?

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -6,12 +6,15 @@ import org.broadinstitute.dsde.rawls.model.Workspace
...
 import spray.http.StatusCodes
+import spray.httpx.SprayJsonSupport._
+import org.broadinstitute.dsde.rawls.model.WorkspaceJsonSupport._
 
 /**
  * Created by dvoet on 4/27/15.
  */
 object WorkspaceService {
   case class SaveWorkspace(workspace: Workspace)
+  case class ListWorkspaces()
```
Comment (mcovarr) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: case classes with no args are deprecated, can this be a case object?

File: src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala [hunk 4]
```diff
@@ -36,20 +33,31 @@ trait PerRequest extends Actor with Json4sSupport {
...
 
   def receive = {
-    case RequestComplete(statusCode, response) => complete(statusCode, response)
-    case ReceiveTimeout => complete(GatewayTimeout, "Request timeout")
-    case response: AnyRef => complete(OK, response)
+    case RequestCompleteOK_(response, marshaller) => complete(OK, response)(marshaller)
+    case RequestComplete_(response, statusCode, marshaller) => complete(statusCode, response)(marshaller)
+    case RequestCompleteNoContent(statusCode) => complete(statusCode)
+    case ReceiveTimeout => complete(GatewayTimeout)
+    case x => {
```
Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala]: Don't need the braces here

File: src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala [hunk 5]
```diff
@@ -36,20 +33,31 @@ trait PerRequest extends Actor with Json4sSupport {
...
+      system.log.error("Unsupported response message sent to PreRequest actor: " + Option(x).getOrElse("null").toString)
+      complete(InternalServerError)
+    }
+  }
+
+  def complete[T](status: StatusCode, obj: T)(implicit marshaller: ToResponseMarshaller[T]) = {
```
Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala]: Can you add comments on what the difference between the two complete()s are and why one would choose one or the other?

File: src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala [hunk 6]
```diff
@@ -57,9 +65,29 @@ trait PerRequest extends Actor with Json4sSupport {
...
+  private object RequestCompleteOK_ {
+    def unapply[T](requestCompleteOK: RequestCompleteOK[T]) = Some((requestCompleteOK.response, requestCompleteOK.marshaller))
+  }
+
+  private object RequestComplete_ {
+    def unapply[T](requestComplete: RequestComplete[T]) = Some((requestComplete.response, requestComplete.statusCode, requestComplete.marshaller))
+  }
 
   case class WithProps(r: RequestContext, props: Props, message: AnyRef, timeout: Duration) extends PerRequest {
     lazy val target = context.actorOf(props, "foo")
```
Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala]: foo? :)

File: src/main/scala/org/broadinstitute/dsde/rawls/ws/RawlsApiService.scala [hunk 7]
```diff
@@ -70,12 +70,12 @@ trait RootRawlsApiService extends HttpService {
 
 @Api(value = "workspace", description = "APIs for Workspace CRUD", position = 1)
 trait WorkspaceApiService extends HttpService with PerRequestCreator {
-  import spray.httpx.SprayJsonSupport._
-  import org.broadinstitute.dsde.rawls.model.WorkspaceJsonSupport._
+    import spray.httpx.SprayJsonSupport._
+    import org.broadinstitute.dsde.rawls.model.WorkspaceJsonSupport._
```
Comment (mcovarr) [inline src/main/scala/org/broadinstitute/dsde/rawls/ws/RawlsApiService.scala]: indent seems funky

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 8]
```diff
@@ -6,12 +6,15 @@ import org.broadinstitute.dsde.rawls.model.Workspace
...
 import org.broadinstitute.dsde.rawls.ws.PerRequest
 import spray.http.StatusCodes
+import spray.httpx.SprayJsonSupport._
+import org.broadinstitute.dsde.rawls.model.WorkspaceJsonSupport._
 
 /**
  * Created by dvoet on 4/27/15.
  */
 object WorkspaceService {
   case class SaveWorkspace(workspace: Workspace)
```
Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: I like defining a sealed trait to define the concept of "message for this actor" and then extending that w/ all my message class/objects

    sealed trait WorkspaceServiceMessage
    case class SaveWorkspace(workspace: Workspace) extends WorkspaceServiceMessage

It makes it clear that I'm looking at an actor message

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/WorkspaceDAO.scala (hunk 1 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/WorkspaceDAO.scala]: yeah, but the java.nio api kind of forces my hand. This is how it is done if we want to use nio.

File: src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala (hunk 6 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala]: gotta problem with foo? and you call yourself a programmer...

Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala]: bar

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/WorkspaceDAO.scala (hunk 1 above)
Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/WorkspaceDAO.scala]: that's what made me a sad panda :(

File: src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala (hunk 6 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/ws/PerRequest.scala]: Looks like we've got a couple of foo fighters here.
